### PR TITLE
Promote PodDisruptionBudget e2e test to Conformance

### PR DESCRIPTION
**Pull request by mortent**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

PDBs are graduating to GA (https://github.com/kubernetes/kubernetes/pull/99290). We need conformance tests for GA.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Must merge after https://github.com/kubernetes/kubernetes/pull/99290

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
KEP: https://github.com/kubernetes/enhancements/pull/904
/area conformance
@kubernetes/sig-architecture-pr-reviews @kubernetes/sig-apps-pr-reviews @kubernetes/cncf-conformance-wg

/cc @liggitt @johnbelamaric 
/priority important-soon

### CODE REVIEW

**Comment by liggitt** [comment]
For reference, here are the beta PDB endpoints and their e2e test coverage:

https://apisnoop.cncf.io/?zoomed=category-beta-policy

Endpoint | Tested
---------|-------
patchPolicyV1beta1NamespacedPodDisruptionBudgetStatus | Untested
readPolicyV1beta1NamespacedPodDisruptionBudgetStatus | Untested
listPolicyV1beta1PodDisruptionBudgetForAllNamespaces | Untested
deletePolicyV1beta1CollectionNamespacedPodDisruptionBudget | Untested
patchPolicyV1beta1NamespacedPodDisruptionBudget | Untested
listPolicyV1beta1NamespacedPodDisruptionBudget | Untested
replacePolicyV1beta1NamespacedPodDisruptionBudgetStatus | Untested
replacePolicyV1beta1NamespacedPodDisruptionBudget | Tested
deletePolicyV1beta1NamespacedPodDisruptionBudget | Tested
createPolicyV1beta1NamespacedPodDisruptionBudget | Tested
readPolicyV1beta1NamespacedPodDisruptionBudget | Tested

Getting regular e2e tests in place for the untested operations is ideal. Then, as part of the GA API promotion, switch the tests to using the v1 API, then promote to conformance.

**Comment by nan-yu** [comment]
More e2e tests are added in PR #85819 except below related with PDBstatus

- [ ] patchPolicyV1beta1NamespacedPodDisruptionBudgetStatus
- [ ] readPolicyV1beta1NamespacedPodDisruptionBudgetStatus
- [ ] replacePolicyV1beta1NamespacedPodDisruptionBudgetStatus

Adding test coverage for them requires adding client-side code. I guess we don't need to add code only for testing purpose. Is that okay to leave them untested when we go GA? @liggitt

**Comment by liggitt** [comment]
> I guess we don't need to add code only for testing purpose. Is that okay to leave them untested when we go GA?

The intent is for every endpoint/operation for non-optional GA APIs to be tested by a conformance test. You can use the generated client's `UpdateStatus` and `Patch(..., "status")` methods to write to the status subresource, and a [dynamic client](https://github.com/kubernetes/client-go/tree/master/dynamic) to get the status subresource.

**Comment by spiffxp** [comment]
Moved back to in progress column as work remains to be done

**Comment by hh** [comment]
@mortent Let me know if I can give a hand to help move this forward!

**Comment by tnozicka** [comment]
/remove-lifecycle rotten
/reopen

@mortent are you planning to continue with https://github.com/kubernetes/kubernetes/pull/81571 ?

**Comment by mortent** [comment]
@tnozicka I hope so, but I haven't had much time to focus on this lately. The issues that needs to be resolved before we can take PDBs to GA are tracked in https://github.com/orgs/kubernetes/projects/36.

File: test/e2e/apps/disruption.go [hunk 1]
```diff
@@ -58,7 +58,13 @@ var _ = SIGDescribe("DisruptionController", func() {
 		createPDBMinAvailableOrDie(cs, ns, intstr.FromString("1%"))
 	})
 
-	ginkgo.It("should update PodDisruptionBudget status", func() {
+	/*
+	   Release : v1.17
```
Comment (dims) [inline test/e2e/apps/disruption.go:62]: need to bump this right?

File: test/e2e/apps/disruption.go [hunk 2]
```diff
@@ -197,7 +203,13 @@ var _ = SIGDescribe("DisruptionController", func() {
 		})
 	}
 
-	ginkgo.It("should block an eviction until the PDB is updated to allow it", func() {
+	/*
+	   Release : v1.17
```
Comment (dims) [inline test/e2e/apps/disruption.go:207]: need to bump this right?

File: test/e2e/apps/disruption.go (hunk 1 above)
Comment (mortent) [reply test/e2e/apps/disruption.go]: Yeah, I have updated release version to 1.21.

File: test/e2e/apps/disruption.go (hunk 2 above)
Comment (mortent) [reply test/e2e/apps/disruption.go]: Updated.

**Comment by Riaankl** [comment]
From what I can pick up in APISnoop the following `DisruptionBudget` endpoints are touch by the "should observe PodDisruptionBudget status updated" test.
```
patchPolicyV1beta1NamespacedPodDisruptionBudget
readPolicyV1beta1NamespacedPodDisruptionBudget
readPolicyV1beta1NamespacedPodDisruptionBudgetStatus
replacePolicyV1beta1NamespacedPodDisruptionBudget
createPolicyV1beta1NamespacedPodDisruptionBudget
deletePolicyV1beta1NamespacedPodDisruptionBudget
```

**Comment by Riaankl** [comment]
TIme out that seem related to the Promotion
```
• Failure [600.407 seconds]
[sig-apps] DisruptionController
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/apps/framework.go:23
  should block an eviction until the PDB is updated to allow it [Conformance] [It]
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:630
  Mar  8 07:08:27.533: Waiting for the pdb to be processed in namespace disruption-7423
      
  Unexpected error:
      <*errors.errorString | 0xc00024a250>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
      ```
```
Mar  8 07:11:20.305: FAIL: Waiting for the pdb to be processed in namespace disruption-6076
Unexpected error:
    <*errors.errorString | 0xc00024a250>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
```

```
• Failure [600.197 seconds]
[sig-apps] DisruptionController
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/apps/framework.go:23
  should observe PodDisruptionBudget status updated [Conformance] [It]
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:630
  Mar  8 07:11:20.305: Waiting for the pdb to be processed in namespace disruption-6076
      
  Unexpected error:
      <*errors.errorString | 0xc00024a250>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
```

**Comment by spiffxp** [comment]
/milestone v1.21
Will look later, but this is subject to Test Freeze, not Code Freeze

**Comment by liggitt** [comment]
looks like it was an oversight to not pick up the other CRUD e2e tests, opened https://github.com/kubernetes/kubernetes/pull/100092

**Comment by Riaankl** [comment]
Thank you @liggitt I will make sure APISnoop update once #100092 merge and give feedback.

**Comment by Riaankl** [comment]
@liggitt  There are two PDB endpoints that show up as having e2e test in [APISnoop](https://apisnoop.cncf.io/conformance-progress/endpoints/1.21.0/?filter=promoted-without-tests) but not conformance tests:

- `replacePolicyV1NamespacedPodDisruptionBudgetStatus` 
- `patchPolicyV1NamespacedPodDisruptionBudget `
The test [`[sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go#L296-L361) , touch both these endpoints but the test have not been promoted to conformance.

**Comment by Riaankl** [comment]
You are correct, `replacePolicyV1NamespacedPodDisruptionBudgetStatus` is conformance tested, it was a typing error on my side. `replacePolicyV1NamespacedPodDisruptionBudget` is the endpoint I should have referred to. 

Both `patch- and replace- PolicyV1NamespacedPodDisruptionBudget` are hit by [`[sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go#L296-L361) 

Are both these dependant on `/eviction` ?
Will they be Promoted to conformance in 1.22?

**Comment by liggitt** [comment]
> Both `patch- and replace- PolicyV1NamespacedPodDisruptionBudget` are hit by [`[sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go#L296-L361)
> 
> Are both these dependant on `/eviction` ?

Those endpoints are unrelated to `/eviction`, they just happened to be tested in that test.

> Will they be Promoted to conformance in 1.22?

The intent is to promote eviction to v1 in 1.22 along with that test, yes.

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/100260 to replicate the patch and update API calls from the eviction test into a CRUD conformance test

**Comment by hh** [comment]
Thanks @liggitt

**Comment by hh** [comment]
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements

> tests may need to create or set objects or fields that are alpha or beta that bypass policies that are not yet GA, but which may reasonably be enabled on a conformant cluster (e.g., pod security policy, non-GA scheduler annotations)

Does this part of the policy allow us to promote the existing test?

**Comment by liggitt** [comment]
> > tests may need to create or set objects or fields that are alpha or beta that bypass policies that are not yet GA, but which may reasonably be enabled on a conformant cluster (e.g., pod security policy, non-GA scheduler annotations)
> 
> Does this part of the policy allow us to promote the existing test?

I would not expect so. My reading of that is that those should be _incidental_ to the test and failure should be tolerated (e.g. attempting to grant a privileged pod security policy so that a privileged test can run).

**Comment by johnbelamaric** [comment]
Agree with Jordan here. That is intended to, for example, allow the test to
create RBAC resources (though it should still tolerate RBAC not being
enabled). It's not intended to allow things that are core to the test to be
non-GA.

I haven't looked at the test, but presumably you could split it up so that
everything is covered in 1.21. We have a little time before the test
freeze, and we have relaxed the flakiness duration in other cases. If you
can make it happen quickly and it's testing things that were non-flaky in
the current version of the test, we should be OK.

**Pull request merged**